sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,276 @@
+# Shared DPC++ libraries
+
+This document describes purpose and design of Shared DPC++ libraries feature.
+
+## Background
+Sometimes users want to provide *device* functions via shared libraries.
```

Comment:
Suggest to say "Sometimes users want to link device code dynamically at run time."

I think we can model a situation where there are no shared libraries (.so/.dll) in the app, just executable, but still dynamic device code linkage is used.

Suppose we have `SYCL_EXTERNAL foo();` called from a kernel but. SYCL app can only define host variant of the function, and dynamically linked device implementation of the function is provided in a separate device binary linked via existing driver options:

a.cpp:
```cpp
SYCL_EXTERNAL void foo();
...
parallel_for([]() {  foo(); });
```
b.cpp:
```cpp
/*no SYCL_EXTERNAL*/ void foo() { ... }
```
Then user creates (or gets somewhere) a SPIRV with `foo` implementation and uses `-fsycl-add-targets` to add it to the executable.

This is not mainstream scenario, but I think design should still work in this case.

Reply:
I think it is possible to support this scenario with current design. However right now `fsycl-add-targets` option passes binaries to `clang-offload-wrapper` bypassing `sycl-post-link` tool.
If we will be able to run SPIR-V binaries from `-fsycl-add-targets` through `sycl-post-link` so it generates lists of exports and inputs for this binary, then it should work. @mdtoguchi , @AGindinson do you think it is possible to make such change in the driver?

Reply:
> If we will be able to run SPIR-V binaries from -fsycl-add-targets through sycl-post-link so it generates lists of exports and inputs for this binary, then it should work.

How would the tool gather the information about symbols based on the SPIR-V image alone? Sorry if I'm missing out on some key concepts that I haven't acquainted myself with yet.

I understand that the driver change itself should be fairly straightforward - just adjusting the action graph for the `-fsycl-add-targets` flow.

Reply:
> How would the tool gather the information about symbols based on the SPIR-V image alone? Sorry if I'm missing out on some key concepts that I haven't acquainted myself with yet.

Yes, we cannot collect symbols information based only on SPIR-V image in `sycl-post-link` unless we translate it back to LLVM IR then back to SPIR-V after we collected symbols info, it doesn't seem like an optimal solution, but I guess it is still 
possible to implement, right?

Another way is to try to pass some kind of saved info along with SPIR-V image through `-fsycl-add-targets` but it will require additional work from user's side and we need somehow add possibility to save symbols info on disk, so it seems not really user-friendly.

Reply:
It is a good point about the symbol table. There can be other kind of info which has to be attached - spec constants, dead args. Users must provide it if they want to use -fsycl-add-targets. I'm not sure if this option works today in all possible cases - probably some fixes would be needed.
There is another way to add an extra device binary - wrap all the needed files and the device binary into an object using clang-offload-wrapper and then link the SYCL app with the object, -fsycl-add-targets is supposed to be  a user-friendly shortcut to this operation. Would be actually good to have a test which attaches device dynamic lib this way.

My main point is that the design should not assume "device dynamic lib" - "host shared lib" 1-1 correspondence, as this would greatly limit flexibility of device code packaging/management. Unless there are some serious technical hurdles impractical to try to solve. But looks like this is not the case, looking at the discussion above.

Reply:
To be honest, I'm failing to deduce why `-fsycl-add-targets` would be needed, at least from the above-demonstrated use-cases.
I believe that the correct way to produce a re-usable library would be to use `-shared -fPIC` (this would give a fat .so library) or `-fsycl-link` (this would give a device-only library). In both cases, the device image would end up wrapped together with all the properties that our flow has assigned to it during compile-time, including the exported/imported symbols' list. It's also known today that `-shared` mechanism is an out-of-the-box one - I've checked that a simple app can be successfully linked with a simple `SYCL_EXTERNAL` function that has been compiled into such a library (it can even be executed, due to its simplicity). `-fsycl-link` might require some additional work if someone would find it necessary to produce "strictly device-only" libraries, but the point stands. As long as we have intuitive ways of producing shared objects with SYCL specifics, and then linking apps against them by listing the new libraries as mere inputs to the link command, we do not need to consider complicated use-cases with manual, error-prone editing on the user side.

Reply:
I removed binding to shared libraries and added mentioned scenario.